File: greedy/swift/665.non-decreasing-array.swift
/*
 * @lc app=leetcode id=665 lang=swift
 *
 * [665] Non-decreasing Array
 */

// @lc code=start
class Solution {
    func checkPossibility(_ nums: [Int]) -> Bool {
        var nums_copy = nums
        var count = 0
        for i in 1..<nums_copy.count {
            if nums_copy[i] < nums_copy[i - 1] {
                count += 1

                if i - 2 >= 0 && nums_copy[i] < nums_copy[i - 2] {
                    nums_copy[i] = nums_copy[i - 1]
                } else {
                    nums_copy[i - 1] = nums_copy[i]
                }
            }
        }
        return count <= 1
    }
}
// @lc code=end

File: greedy/cpp/665.non-decreasing-array.cpp
/*
 * @lc app=leetcode id=665 lang=cpp
 *
 * [665] Non-decreasing Array
 */

// @lc code=start
#include <vector>
using namespace std;

class Solution {
public:
    bool checkPossibility(vector<int>& nums) {
        int count = 0;
        for (int i = 1; i < nums.size(); ++i) {
            if (nums[i] < nums[i - 1]) {
                ++count;
                if (count > 1) return false;

                if (i == 1 || nums[i - 2] <= nums[i]) {
                    nums[i - 1] = nums[i];
                } else {
                    nums[i] = nums[i - 1];
                }
            }
        }

        return count <= 1;

    }
};
// @lc code=end

